clients/src/main/java/org/apache/kafka/common/protocol/Readable.java

```diff
@@ -35,6 +36,10 @@
     int readUnsignedVarint();
     ByteBuffer readByteBuffer(int length);
 
+    default BaseRecords readRecords(int length) {
+        throw new UnsupportedOperationException("Not implemented");
+    }
```

Comment:
Is this and the respective `writeRecords` in the base interface needed? It seems like they're only implemented in two specific cases. Could we not downcast for those cases?